packages/react/src/components/Geo/Geocoder/index.tsx

```diff
@@ -0,0 +1,243 @@
+import maplibregl from 'maplibre-gl';
+import { createAmplifyGeocoder } from 'maplibre-gl-js-amplify';
+import React, { useEffect } from 'react';
+import { useControl, useMap } from 'react-map-gl';
+import type { IControl } from 'react-map-gl';
+
+const GEOCODER_OPTIONS = {
+  maplibregl,
+  marker: { color: '#3FB1CE' },
+  popup: true,
+  showResultMarkers: { color: '#3FB1CE' },
+  showResultsWhileTyping: true,
+};
+
+const GEOCODER_CONTAINER = 'geocoder-container';
+
+type GeocoderControl = IControl & {
+  addTo: (container: string) => void;
+};
+
+/**
+ * The `<Geocoder>` component provides location search. See https://ui.docs.amplify.aws/components/geo#geocoder.
+ *
+ * @example
+ * // Used as a map control:
+ * function App() {
+ *   return (
+ *     <AmplifyMap>
+ *       <Geocoder />
+ *     </AmplifyMap>
+ *   );
+ * }
+ *
+ * @example
+ * // Used as a standalone component:
+ * function App() {
+ *   return <Geocoder />;
+ * }
+ */
+export const Geocoder = ({
+  position = 'top-right',
+  ...props
+}: GeocoderProps) => {
+  const { current: map } = useMap();
+
+  /**
+   * This logic determines whether the Geocoder exists as part of a Map component or if it is a standalone component.
+   * The `useControl` hook from `react-map-gl` makes it easy to add a control to a map, but throws an error if that map
+   * doesn't exist. If the map doesn't exist, the Geocoder is mounted to a container upon rendering.
+   */
+  if (map) {
+    useControl(
+      () =>
+        createAmplifyGeocoder({
+          ...GEOCODER_OPTIONS,
+          ...props,
+        }) as unknown as GeocoderControl
+    );
+  } else {
+    useEffect(() => {
+      (
+        createAmplifyGeocoder({
+          ...GEOCODER_OPTIONS,
+          ...props,
+        }) as unknown as GeocoderControl
+      ).addTo(`#${GEOCODER_CONTAINER}`);
+    }, []);
+  }
```

Comment:
Hi, sorry -- didn't get the time to review it yesterday. Technically this breaks the first rule of the hooks: ([overview](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level), [details](https://reactjs.org/docs/hooks-rules.html#explanation)).

Have you looked into this a little? While this could be okay because React will be oblivious as long as the number of hooks is consistent, this might eventually bite us later. cc @calebpollman as he's more knowledgable about hooks

technically